Mage/src/main/java/mage/abilities/common/CastAsThoughItHadFlashIfConditionAbility.java

```diff
@@ -0,0 +1,81 @@
+package mage.abilities.common;
+
+import mage.abilities.Ability;
+import mage.abilities.condition.Condition;
+import mage.abilities.effects.ContinuousRuleModifyingEffectImpl;
+import mage.abilities.effects.common.continuous.CastAsThoughItHadFlashSourceEffect;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.constants.Zone;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+
+/**
+ *
+ * @author awjackson
+ */
+public class CastAsThoughItHadFlashIfConditionAbility extends SimpleStaticAbility {
+
+    private final String rule;
+
+    public CastAsThoughItHadFlashIfConditionAbility(Condition condition, String rule) {
+        super(Zone.ALL, new CastAsThoughItHadFlashSourceEffect(Duration.EndOfGame));
+        this.addEffect(new CastAsThoughItHadFlashIfConditionEffect(condition));
+        this.setRuleAtTheTop(true);
+        this.rule = rule;
+    }
+
+    private CastAsThoughItHadFlashIfConditionAbility(final CastAsThoughItHadFlashIfConditionAbility ability) {
+        super(ability);
+        this.rule = ability.rule;
+    }
+
+    @Override
+    public CastAsThoughItHadFlashIfConditionAbility copy() {
+        return new CastAsThoughItHadFlashIfConditionAbility(this);
+    }
+
+    @Override
+    public String getRule() {
+        return rule;
+    }
+}
+
+class CastAsThoughItHadFlashIfConditionEffect extends ContinuousRuleModifyingEffectImpl {
+
+    private final Condition condition;
+
+    public CastAsThoughItHadFlashIfConditionEffect(Condition condition) {
+        super(Duration.EndOfGame, Outcome.Neutral);
+        this.condition = condition;
+    }
+
+    private CastAsThoughItHadFlashIfConditionEffect(final CastAsThoughItHadFlashIfConditionEffect effect) {
+        super(effect);
+        this.condition = effect.condition;
+    }
+
+    @Override
+    public CastAsThoughItHadFlashIfConditionEffect copy() {
+        return new CastAsThoughItHadFlashIfConditionEffect(this);
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.CAST_SPELL_LATE;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        if (!event.getSourceId().equals(source.getSourceId())) {
+            return false;
+        }
+        if (game.isMainPhase() && game.isActivePlayer(event.getPlayerId()) && game.getStack().size() == 1) {
+            return false;
+        }
+        if (game.inCheckPlayableState()) {
+            return false;
+        }
```

Comment:
I do not understand these checks at all

Reply:
The first check is basically `game.canPlaySorcery()` except that method checks that the stack is empty.  Here the spell has already been put on the stack so it checks that stack size is 1 instead.

`inCheckPlayableState` is how xmage highlights cards that are castable (and allows casting them at all).  Player actions (like selecting the X cost) can't be known when looking at would *could* be castable so we just err on the side of making it castable for the purpose of those checks.

Reply:
Why is it returning false for all of these checks?

Reply:
`ContinuousRuleModifyingEffect.applies` returns `true` if the event being checked should be prohibited by the effect, and `false if the event is not prohibited.

Reply:
> `ContinuousRuleModifyingEffect.applies` returns `true` if the event being checked should be prohibited by the effect, and `false` if the event is not prohibited.

Boy that's confusing.

Reply:
It is called "applies" (i.e. does the prohibition *apply* to the proposed event) and not something like "allowed".

Reply:
> It is called "applies" (i.e. does the prohibition _apply_ to the proposed event) and not something like "allowed".

I find it confusing given the class name `CastAsThoughItHadFlashIfConditionEffect`. `CastAsThoughItHadFlashIfConditionEffect.applies == true` based on the naming convention for the rest of the code would lead me to believe that the spell *can* be cash as though it had flash.

Reply:
Good point, I will rename the class. I was following the general naming convention that an effect that is only used by [Foo]Ability is named [Foo]Effect unless an ability has multiple unique effects, but in this case the name is confusing.